File: frontend/src/routes/_shared/reports.ts
import { requestJson } from './api';

export type ReportCommunitySummary = {
  community_id?: number | null;
  human_readable_id?: number | null;
  parent?: number | null;
  children?: number[] | null;
  title?: string | null;
  summary?: string | null;
  findings?: unknown;
  rating?: number | null;
  size?: number | null;
};

export type ReportCommunityListResponse = {
  collection_id: string;
  level?: number | null;
  total: number;
  count: number;
  items: ReportCommunitySummary[];
};

export type ReportCommunityDetailResponse = ReportCommunitySummary & {
  collection_id: string;
  level?: number | null;
  document_count?: number | null;
  text_unit_count?: number | null;
  entities: Array<Record<string, unknown>>;
  relationships: Array<Record<string, unknown>>;
  documents: Array<Record<string, unknown>>;
};

export type ReportPatternItem = {
  community_id?: number | null;
  title?: string | null;
  summary?: string | null;
  findings?: unknown;
  rating?: number | null;
  size?: number | null;
  level?: number | null;
};

export type ReportPatternsResponse = {
  collection_id: string;
  level?: number | null;
  total_communities: number;
  total_entities?: number | null;
  total_relationships?: number | null;
  total_documents?: number | null;
  count: number;
  items: ReportPatternItem[];
};

export async function listCommunityReports(
  collectionId: string,
  options: { level?: number; limit?: number; offset?: number; minSize?: number; sort?: string } = {}
) {
  const params = new URLSearchParams({
    level: String(options.level ?? 2),
    limit: String(options.limit ?? 20),
    offset: String(options.offset ?? 0),
    min_size: String(options.minSize ?? 0),
    sort: options.sort ?? 'rating'
  });

  return (await requestJson(
    `/collections/${encodeURIComponent(collectionId)}/reports/communities?${params.toString()}`,
    { method: 'GET' }
  )) as ReportCommunityListResponse;
}

export async function getCommunityReportDetail(
  collectionId: string,
  communityId: string,
  options: { level?: number; entityLimit?: number; relationshipLimit?: number; documentLimit?: number } = {}
) {
  const params = new URLSearchParams();
  if (typeof options.level === 'number') params.set('level', String(options.level));
  params.set('entity_limit', String(options.entityLimit ?? 12));
  params.set('relationship_limit', String(options.relationshipLimit ?? 12));
  params.set('document_limit', String(options.documentLimit ?? 12));

  return (await requestJson(
    `/collections/${encodeURIComponent(collectionId)}/reports/communities/${encodeURIComponent(
      communityId
    )}?${params.toString()}`,
    { method: 'GET' }
  )) as ReportCommunityDetailResponse;
}

export async function listReportPatterns(
  collectionId: string,
  options: { level?: number; limit?: number; sort?: string } = {}
) {
  const params = new URLSearchParams({
    level: String(options.level ?? 2),
    limit: String(options.limit ?? 6),
    sort: options.sort ?? 'rating'
  });

  return (await requestJson(
    `/collections/${encodeURIComponent(collectionId)}/reports/patterns?${params.toString()}`,
    { method: 'GET' }
  )) as ReportPatternsResponse;
}

File: backend/controllers/reports.py
from __future__ import annotations

from fastapi import APIRouter, Query

from application import report_service
from controllers.schemas.report import (
    ReportCommunityDetailResponse,
    ReportCommunityListResponse,
    ReportPatternsResponse,
)

router = APIRouter(tags=["reports"])


@router.get(
    "/collections/{collection_id}/reports/communities",
    response_model=ReportCommunityListResponse,
    summary="列出社区报告",
)
async def list_community_reports(
    collection_id: str,
    level: int | None = Query(default=2, description="社区层级"),
    limit: int = Query(default=50, ge=1, le=200, description="返回数量"),
    offset: int = Query(default=0, ge=0, description="偏移量"),
    min_size: int = Query(default=0, ge=0, description="最小社区规模过滤"),
    sort: str | None = Query(default="rating", description="排序字段（rating/size）"),
) -> ReportCommunityListResponse:
    return report_service.list_community_reports(
        collection_id=collection_id,
        level=level,
        limit=limit,
        offset=offset,
        min_size=min_size,
        sort=sort,
    )


@router.get(
    "/collections/{collection_id}/reports/communities/{community_id}",
    response_model=ReportCommunityDetailResponse,
    summary="社区报告详情",
)
async def get_community_report_detail(
    collection_id: str,
    community_id: str,
    level: int | None = Query(default=None, description="指定社区层级"),
    entity_limit: int = Query(default=20, ge=1, le=200, description="实体返回数量"),
    relationship_limit: int = Query(
        default=20, ge=1, le=200, description="关系返回数量"
    ),
    document_limit: int = Query(default=20, ge=1, le=200, description="文档返回数量"),
) -> ReportCommunityDetailResponse:
    return report_service.get_community_report_detail(
        collection_id=collection_id,
        community_id=community_id,
        level=level,
        entity_limit=entity_limit,
        relationship_limit=relationship_limit,
        document_limit=document_limit,
    )


@router.get(
    "/collections/{collection_id}/reports/patterns",
    response_model=ReportPatternsResponse,
    summary="社区规律概览",
)
async def list_patterns(
    collection_id: str,
    level: int | None = Query(default=2, description="社区层级"),
    limit: int = Query(default=10, ge=1, le=50, description="返回数量"),
    sort: str | None = Query(default="rating", description="排序字段（rating/size）"),
) -> ReportPatternsResponse:
    return report_service.list_patterns(
        collection_id=collection_id,
        level=level,
        limit=limit,
        sort=sort,
    )
